### PR TITLE
fix(classification): enforce confidence threshold for domain signal matching

### LINKED ISSUE #952: bug: Two prompts with same semantics but different spelling/grammar, get different classification results
### Describe the bug

```
(base) ➜  ~ curl http://localhost:8080/api/v1/classify/intent \
  -H "Content-Type: application/json" \
  -d '{
    "text": "Please solve the following mathematical problem step by step. \n\nContext: The perimeter of two squares ABCD and PQRS are 3:2 respectively. The sums of the perimeters of the squares are 30.\n\nQuestion: Find the perimeter of ABCD and PQRS.\n\nProvide your final answer in \\boxed{{}} format, where the content inside the braces is the exact mathematical expression or number. For example: \\boxed{{42}}. Keep your explanation clear, concise, and within 3 sentences."
  }'
{"classification":{"category":"math","confidence":0.7054473161697388,"processing_time_ms":362},"recommended_model":"math-specialized-model","routing_decision":"high_confidence_specialized"}

(base) ➜  ~ curl http://localhost:8080/api/v1/classify/intent \
  -H "Content-Type: application/json" \
  -d '{
    "text": "Plese slove th fllowing mathemtical prblem stp by stp.\n\nContxt: Th perimetr of two sqaures ABCD n PQRS ar 3:2 respctivly. The sum of th prmeters of th sqares ar 30.\n\nQuestin: Fnd the perimeter of ABCD nd PQRS.\n\nProvde yur fnal answr in \\boxed{{}} frmat, whre the contnt insde the bracss is the xact mathemtical exprssion or numbr. Fr exampl: \\boxed{{42}}. Kep yur explanaton cleer, cnsise, nd wthin 3 sentnces."
  }'

{"classification":{"category":"chemistry","confidence":0.43069133162498474,"processing_time_ms":504},"recommended_model":"chemistry-specialized-model","routing_decision":"low_confidence_general"}
```

### To Reproduce

```
diff --git a/src/vllm-sr/cli/docker_cli.py b/src/vllm-sr/cli/docker_cli.py
index e2c9dbe..bed122e 100644
--- a/src/vllm-sr/cli/docker_cli.py
+++ b/src/vllm-sr/cli/docker_cli.py
@@ -268,6 +268,8 @@ def docker_start_vllm_sr(
             "50051:50051",  # Router gRPC port (internal)
             "-p",
             "9190:9190",  # Metrics port
+            "-p",
+            "8080:8080",  # Router API port
             # Note: 8080 (Router API) is not exposed by default
             # Health checks are done inside the container via docker exec
         ]
```
please setup a router intance with vllm-sr, then reproduce this issue via calling the API endpoint /api/v1/classify/intent with above two prompts, and confirm that they get different classification results.

### Expected behavior

Get same category

### Additional context

[robustness_flips.csv](https://github.com/user-attachments/files/24428514/robustness_flips.csv) is generated via https://github.com/carlory/RouterArena/blob/pr-vllm-deepseek/scripts/analyze_robustness_flips.py and https://github.com/carlory/RouterArena/blob/pr-vllm-deepseek/scripts/analyze_prompt_diffs.py

### PR DESCRIPTION
**Pull request by asaadbalum**
## Summary

Fixes incorrect classification of typo-laden text by enforcing confidence threshold checks before domain signals are used in routing decisions. Low-confidence ML model predictions were previously still matching decisions, causing misspelled prompts to be routed incorrectly.

## Related Issue

Fixes #952

## Problem

When prompts contain typos or misspellings, the ML classification model may produce low-confidence predictions. However, the decision engine was using these low-confidence classifications to match routing decisions without checking against the configured threshold. This caused typo-laden text to be incorrectly routed (e.g., a misspelled math problem classified as "chemistry" instead of "math").

Example from issue:
- Clean prompt: "Please solve this math problem" → correctly classified as "math" (confidence: 0.71)
- Typo prompt: "Plese slove this math prblem" → incorrectly classified as "chemistry" (confidence: 0.43, below threshold of 0.6)

## Solution

1. **Confidence Threshold Enforcement** (`pkg/classification/classifier.go`):
   - Added check to only add domain signals to `MatchedDomainRules` when ML model confidence meets the configured threshold (`classifier.category_model.threshold`, default 0.6)
   - Prevents low-confidence misclassifications from triggering incorrect routing decisions

2. **Graceful Fallback** (`pkg/services/classification.go`):
   - When classification fails (e.g., all signals skipped due to low confidence), fall back to "other" category with 0.0 confidence instead of returning an error
   - Ensures the router always returns a valid response even for low-confidence inputs

## Changes Made

| File | Change Type | Description |
|------|-------------|-------------|
| `src/semantic-router/pkg/classification/classifier.go` | Modified | Added confidence threshold check before adding domain signals |
| `src/semantic-router/pkg/services/classification.go` | Modified | Added graceful fallback to "other" category on classification failure |

## Testing Done

- [x] Code compiles successfully (`go build`)
- [x] Go vet passes (`go vet`)
- [x] No linter errors
- [x] Manual verification: Typo prompts now fall back to "other" or are correctly classified instead of being misclassified

### Test Results

**Before Fix:**
- Clean prompt: "Please solve this math problem" → `math` (confidence: 0.71) ✅
- Typo prompt: "Plese slove this math prblem" → `chemistry` (confidence: 0.43) ❌

**After Fix:**
- Clean prompt: "Please solve this math problem" → `math` (confidence: 0.71) ✅
- Typo prompt: "Plese slove this math prblem" → `other` (confidence: 0.0) ✅ (graceful fallback)

### New Retrained Model Available

A retrained model with typo augmentation is available that further improves typo robustness:

**Performance (from GPU testing on 100 samples):**
- **Clean Accuracy**: 70% → 73% (+3%) ✅
- **Typo Accuracy**: 34% → 45% (+11%) ✅

**Model Location:**
- HuggingFace: https://huggingface.co/abalum/mom-domain-classifier-v1-typo-robust
- **To be moved to:** `LLM-Semantic-Router/lora_intent_classifier_bert-base-uncased_model-v1-typo-robust` (with version tag to preserve original)

### Questions for Reviewers

1. **Model Update Process:**
   - Should this replace the existing model or be a versioned model (e.g., `-v1-typo-robust`)?
   - What's the preferred naming convention for model versions in the org?

2. **Follow-up Enhancement:**
   - Propose opening a new enhancement issue for further typo robustness improvements
   - Will require more investigation and training time
   - Also includes e2e test case for typo robustness (as requested in PR comments)

**Note:** The code fix works with the existing model. The new model provides additional typo robustness but is optional.

## Breaking Changes

None. This is a bug fix that improves robustness without changing the API or configuration format.

## Checklist

- [x] Code follows project style guidelines
- [x] Tests pass (code compiles, vet passes)
- [x] Documentation updated
- [x] Signed-off-by included in commits (DCO)
- [x] No security issues introduced
- [x] No breaking changes

### CODE REVIEW

**Comment by Xunzhuo** [comment]
have you tried this version and make sure the typo issue fixed?

**Comment by Xunzhuo** [comment]
also we need an e2e testcase for this, to make sure the future changes doesnt affect the logics, this can be a follow up

**Comment by asaadbalum** [comment]
Hi @Xunzhuo,

Yes, tested and verified with issue #952 prompts. The entropy threshold enforcement prevents low-confidence misclassifications.

**New Retrained Model:**

Tested new model shows +11% typo accuracy (34% → 45%) and +3% clean accuracy (70% → 73%) from testing on 100 samples.

**Model Integration Questions:**

If accepted, need clarification on:
1. **Naming**: `LLM-Semantic-Router/lora_intent_classifier_bert-base-uncased_model-v1-typo-robust` (versioned) or replace?
2. **Transfer**: How to move from my hg repo `abalum/mom-domain-classifier-v1-typo-robust` to org repo?

And e2e tests will be implemented then.

BTW, I'd suggest to open an enhancement issue, to further enhance the accuracy with misspelled prompts, which will need much more investigation and training time.

**Comment by rootfs** [comment]
@asaadbalum can you add the training script to the repo too? The models will be retrained to support multilingual

**Comment by asaadbalum** [comment]
> @asaadbalum can you add the training script to the repo too? The models will be retrained to support multilingual

@rootfs Added `ft_linear_lora_consistency.py` to the PR.

This script implements **consistency loss training** for typo robustness:
- Creates clean AND typo versions of each sample
- Uses KL divergence to enforce same predictions for both
- Achieved +11% typo accuracy without clean accuracy regression

**Recommended training parameters:**
accelerate launch ft_linear_lora_consistency.py \
    --epochs 15 --max-samples 25000 --typo-prob 0.20 \
    --consistency-weight 1.0 --learning-rate 2e-5 --batch-size 16
    
For multilingual retraining, the augmentation function can be extended to support language-specific typo patterns.

Also, still looking for clarification on my previous questions to @Xunzhuo regarding the retrained model:
1. **Naming**: `lora_intent_classifier_bert-base-uncased_model-v1-typo-robust` (versioned) or replace existing?
2. **Transfer**: How to move from `abalum/mom-domain-classifier-v1-typo-robust` to org repo?
3. Once agreed to use it, will also add a e2e test.
